File: scripts/generate_readme.py
import json
import re
import os

def get_cv_data(script_dir):
    """Reads and parses the cvData object from the javascript file."""
    data_file_path = os.path.join(script_dir, '..', 'assets', 'data', 'data.js')
    try:
        with open(data_file_path, 'r', encoding='utf-8') as f:
            js_content = f.read()
    except FileNotFoundError:
        print(f"Error: {data_file_path} not found.")
        return None

    if 'window.cvData = ' in js_content:
        json_content = js_content.strip().replace('window.cvData = ', '')
        if json_content.endswith(';'):
            json_content = json_content[:-1]
    else:
        print("Error: 'window.cvData' not found in the JavaScript file.")
        return None
        
    try:
        return json.loads(json_content)
    except json.JSONDecodeError as e:
        print(f"Error decoding JSON: {e}")
        return None

def build_markdown_content(full_data, lang_code):
    """Builds the full markdown string for a given language."""
    
    lang_data = full_data.get(lang_code)
    if not lang_data:
        return None

    content = []

    # --- Language Switcher ---
    if lang_code == 'en':
        content.append("[Ver en español](README.es.md)\n")
    elif lang_code == 'es':
        content.append("[Read in English](README.md)\n")

    # --- Profile Section ---
    profile = lang_data.get('profile', {})
    if profile:
        content.append(f"# {profile.get('name', '')}\n")
        content.append(f"## {profile.get('title', '')}\n")
        content.append(f"_{profile.get('tagline', '')}_\n")
        
        contact = profile.get('contact', {})
        contact_links = []
        if contact.get('email'):
            contact_links.append(f"[Email](mailto:{contact.get('email')})")
        if contact.get('linkedin'):
            contact_links.append(f"[LinkedIn]({contact.get('linkedin')})")
        if contact.get('github'):
            contact_links.append(f"[GitHub]({contact.get('github')})")
        if contact.get('youtube'):
            contact_links.append(f"[YouTube]({contact.get('youtube')})")
        if contact.get('location'):
            contact_links.append(f"📍 {contact.get('location')}")
        content.append(" | ".join(contact_links) + "\n")

    # --- About Section ---
    about = lang_data.get('about', {})
    if about:
        title = about.get('title', 'About' if lang_code == 'en' else 'Sobre mí')
        content.append(f"## {title}\n")
        description = re.sub('<[^<]+?>', '', about.get('description', ''))
        content.append(f"{description}\n")

    # --- Experience Section ---
    experience = lang_data.get('experience', [])
    if experience:
        title = 'Experience' if lang_code == 'en' else 'Experiencia'
        content.append(f"## {title}\n")
        for job in experience:
            content.append(f"### {job.get('role', '')} at {job.get('company', '')}\n")
            content.append(f"*{job.get('period', '')}*\n")
            for resp in job.get('responsibilities', []):
                clean_resp = re.sub(r'</?strong>', '**', resp)
                content.append(f"- {clean_resp}")
            content.append("")

    # --- Projects Section ---
    projects = lang_data.get('projects', [])
    if projects:
        title = 'Projects' if lang_code == 'en' else 'Proyectos'
        content.append(f"## {title}\n")
        for proj in projects:
            content.append(f"### {proj.get('name', '')}\n")
            content.append(f"_{proj.get('summary', '')}_\n")
            if proj.get('tech'):
                tech_title = 'Technologies' if lang_code == 'en' else 'Tecnologías'
                content.append(f"**{tech_title}:** {', '.join(proj.get('tech'))}\n")
            content.append(f"{proj.get('description', '')}\n")
    
    # --- Skills Section ---
    skills = lang_data.get('skills', {})
    if skills:
        title = 'Skills' if lang_code == 'en' else 'Habilidades'
        content.append(f"## {title}\n")
        for category, skill_list in skills.items():
            content.append(f"- **{category}:** {', '.join(skill_list)}")
        content.append("\n")

    # --- Education Section ---
    education = lang_data.get('education', [])
    if education:
        title = lang_data.get('education_title', 'Education' if lang_code == 'en' else 'Educación')
        content.append(f"## {title}\n")
        for edu in education:
            content.append(f"### {edu.get('degree', '')}\n")
            content.append(f"*{edu.get('institution', '')} ({edu.get('period', '')})*\n")
            if edu.get('achievements'):
                content.append(f"- {edu.get('achievements')}\n")

    # --- Certifications Section ---
    certifications = lang_data.get('certifications', {})
    if certifications and certifications.get('list'):
        content.append(f"## {certifications.get('title', 'Certifications')}\n")
        for cert_group in certifications.get('list', []):
            content.append(f"### {cert_group.get('name')}")
            for item in cert_group.get('items', []):
                content.append(f"- {item}")
        content.append("\n")
        
    # --- Courses Section ---
    courses = lang_data.get('courses', {})
    if courses and courses.get('list'):
        content.append(f"## {courses.get('title', 'Courses')}\n")
        for course_group in courses.get('list', []):
            content.append(f"### {course_group.get('name')}")
            for item in course_group.get('items', []):
                content.append(f"- {item}")
        content.append("\n")
        
    return "\n".join(content)

def write_readme_file(script_dir, content, filename):
    """Writes content to a file in the root directory."""
    readme_path = os.path.join(script_dir, '..', filename)
    try:
        with open(readme_path, 'w', encoding='utf-8') as f:
            f.write(content)
        print(f"{filename} generated successfully.")
    except IOError as e:
        print(f"Error writing to {filename}: {e}")

def main():
    """Main function to generate README files."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    full_data = get_cv_data(script_dir)
    if not full_data:
        return

    # Generate English README
    en_content = build_markdown_content(full_data, 'en')
    if en_content:
        write_readme_file(script_dir, en_content, 'README.md')

    # Generate Spanish README
    es_content = build_markdown_content(full_data, 'es')
    if es_content:
        write_readme_file(script_dir, es_content, 'README.es.md')

if __name__ == "__main__":
    main()

File: scripts/generate_readme.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"text/template"
)

// Structs to match the JSON structure from data.js
type FullCVData struct {
	EN CVContent `json:"en"`
	ES CVContent `json:"es"`
}

type CVContent struct {
	Profile      Profile             `json:"profile"`
	About        About               `json:"about"`
	Experience   []Experience        `json:"experience"`
	Education    []Education         `json:"education"`
	Certificates Certifications      `json:"certifications"`
	Courses      Courses             `json:"courses"`
	Skills       map[string][]string `json:"skills"`
	Projects     []Project           `json:"projects"`
}

type Profile struct {
	Name     string  `json:"name"`
	Title    string  `json:"title"`
	Tagline  string  `json:"tagline"`
	Contact  Contact `json:"contact"`
}

type Contact struct {
	Email    string `json:"email"`
	LinkedIn string `json:"linkedin"`
	GitHub   string `json:"github"`
	YouTube  string `json:"youtube"`
	Location string `json:"location"`
}

type About struct {
	Title       string `json:"title"`
	Description string `json:"description"`
}

type Experience struct {
	Company         string   `json:"company"`
	Period          string   `json:"period"`
	Role            string   `json:"role"`
	Responsibilities []string `json:"responsibilities"`
}

type Education struct {
	Institution  string `json:"institution"`
	Degree       string `json:"degree"`
	Period       string `json:"period"`
	Achievements string `json:"achievements"`
}

type Certifications struct {
	Title string      `json:"title"`
	List  []NameItems `json:"list"`
}

type Courses struct {
	Title string      `json:"title"`
	List  []NameItems `json:"list"`
}

type NameItems struct {
	Name  string   `json:"name"`
	Items []string `json:"items"`
}

type Project struct {
	Name        string   `json:"name"`
	Summary     string   `json:"summary"`
	Tech        []string `json:"tech"`
	Description string   `json:"description"`
}

// TemplateData holds the content for a specific language and the switcher info
type TemplateData struct {
	Content      CVContent
	SwitcherLink string
	SwitcherText string
}

// Updated Markdown template with a language switcher
const readmeTemplate = `[{{.SwitcherText}}]({{.SwitcherLink}})

# {{.Content.Profile.Name}}

## {{.Content.Profile.Title}}

_{{.Content.Profile.Tagline}}_ 

[Email](mailto:{{.Content.Profile.Contact.Email}}) | [LinkedIn]({{.Content.Profile.Contact.LinkedIn}}) | [GitHub]({{.Content.Profile.Contact.GitHub}}) | [YouTube]({{.Content.Profile.Contact.YouTube}}) | 📍 {{.Content.Profile.Contact.Location}}

## {{.Content.About.Title}}

{{.Content.About.Description | cleanHTML}}

## Experience
{{range .Content.Experience}}
### {{.Role}} at {{.Company}}

*{{.Period}}*
{{range .Responsibilities}}
- {{ . | cleanStrong}}{{end}}
{{end}}

## Projects
{{range .Content.Projects}}
### {{.Name}}

_{{.Summary}}_ 

**Technologies:** {{join .Tech ", "}}

{{.Description}}
{{end}}

## Skills
{{range $category, $skills := .Content.Skills}}
- **{{$category}}:** {{join $skills ", "}}
{{end}}

## Education
{{range .Content.Education}}
### {{.Degree}}

*{{.Institution}} ({{.Period}})*
{{if .Achievements}}
- {{.Achievements}}
{{end}}
{{end}}

## {{.Content.Certificates.Title}}
{{range .Content.Certificates.List}}
### {{.Name}}
{{range .Items}}
- {{.}}
{{end}}{{end}}

## {{.Content.Courses.Title}}
{{range .Content.Courses.List}}
### {{.Name}}
{{range .Items}}
- {{.}}
{{end}}{{end}}
`

// generateReadme creates a README file from the given data and template.
func generateReadme(data TemplateData, tmpl *template.Template, outputFilename string, projectDir string) {
	readmePath := filepath.Join(projectDir, outputFilename)

	// Create the output file.
	file, err := os.Create(readmePath)
	if err != nil {
		log.Fatalf("Error creating file %s: %v", readmePath, err)
	}
	defer file.Close()

	// Execute the template with the data and write the output to the file.
	err = tmpl.Execute(file, data)
	if err != nil {
		log.Fatalf("Error executing template for %s: %v", outputFilename, err)
	}

	fmt.Printf("Successfully generated %s\n", readmePath)
}

func main() {
	// Get the current working directory, which should be the project root.
	projectDir, err := os.Getwd()
	if err != nil {
		log.Fatalf("Error getting current working directory: %v", err)
	}

	// Build path to the data file, relative to the project root.
	dataFilePath := filepath.Join(projectDir, "assets", "data", "data.js")

	// Read the JavaScript file.
	jsContent, err := ioutil.ReadFile(dataFilePath)
	if err != nil {
		log.Fatalf("Error reading %s: %v", dataFilePath, err)
	}

	// Clean the JS content to extract the raw JSON object string.
	contentStr := string(jsContent)
	contentStr = strings.TrimSpace(contentStr)
	contentStr = strings.TrimPrefix(contentStr, "window.cvData = ")
	contentStr = strings.TrimSuffix(contentStr, ";")

	// Unmarshal the JSON string into the Go structs.
	var cvData FullCVData
	err = json.Unmarshal([]byte(contentStr), &cvData)
	if err != nil {
		log.Fatalf("Error unmarshaling JSON: %v", err)
	}

	// Define custom functions for the template.
	funcMap := template.FuncMap{
		"join":      strings.Join,
		"cleanHTML": func(s string) string { return regexp.MustCompile("<[^>]*>").ReplaceAllString(s, "") },
		"cleanStrong": func(s string) string { return regexp.MustCompile(`</?strong>`).ReplaceAllString(s, "**") },
	}

	// Create and parse the template.
	tmpl, err := template.New("readme").Funcs(funcMap).Parse(strings.TrimSpace(readmeTemplate))
	if err != nil {
		log.Fatalf("Error parsing template: %v", err)
	}

	// Generate English README
	enData := TemplateData{
		Content:      cvData.EN,
		SwitcherLink: "README.es.md",
		SwitcherText: "Ver en español",
	}
	generateReadme(enData, tmpl, "README.md", projectDir)

	// Generate Spanish README
	esData := TemplateData{
		Content:      cvData.ES,
		SwitcherLink: "README.md",
		SwitcherText: "Read in English",
	}
	generateReadme(esData, tmpl, "README.es.md", projectDir)
}